src/miral/application_selector.cpp

```diff
@@ -21,243 +21,439 @@
+        {
+            return comparison_func(application, existing_application);
+        });
+    }
+
+    auto advance(bool reverse) -> T
+    {
+        if (focus_list.empty())
+            return T();
+
+        if (!is_active)
+        {
+            is_active = true;
+            originally_selected = focus_list.front();
+            selected = originally_selected;
+        }
+
+        // Attempt to focus the next element after the selected element.
+        auto it = find(selected);
+        auto start_it = it;
+
+        do {
+            if (reverse)
+            {
+                if (it == focus_list.begin())
+                {
+                    it = focus_list.end() - 1;
+                }
+                else
+                {
+                    it--;
+                }
+            }
+            else
+            {
+                it++;
+                if (it == focus_list.end())
+                {
+                    it = focus_list.begin();
+                }
+            }
+
+            // We made it all the way through the list but failed to find anything.
+            if (it == start_it)
+                break;
+        } while (!can_select(*it));
+
+        on_raised(*start_it, *it, originally_selected);
+        return *it;
+    }
+
+    std::function<bool(T const&, T const&)> comparison_func;
+    std::function<bool(T const&)> can_select;
+    std::function<void(T const&, T const&, T const&)> on_raised;
+    std::function<void(T const&)> on_cancelled;
+    std::function<void(T&)> reset;
+
+    std::vector<T> focus_list;
+    /// The application that was selected when next was first called
+    T originally_selected;
+
+    /// The application that is currently selected.
+    T selected;
+
+    bool is_active = false;
+};
+
+}
+
+ApplicationSelector::ApplicationSelector(miral::WindowManagerTools const& _tools) :
+    tools{_tools},
+    window_cycler{std::make_unique<WindowCyclingBehavior<WeakApplication>>(
+        get_comparison_func(),
+        get_can_select_func(tools),
+        get_on_raised_func(tools),
+        get_on_cancelled_func(tools),
+        get_reset_func()
+    )}
+{
+}
 
 ApplicationSelector::~ApplicationSelector() = default;
 
 ApplicationSelector::ApplicationSelector(miral::ApplicationSelector const& other) :
-    tools{other.tools},
-    focus_list{other.focus_list},
-    originally_selected{other.originally_selected},
-    selected{other.selected},
-    active_window{other.active_window}
+    tools{other.tools}
 {
+    window_cycler = std::make_unique<WindowCyclingBehavior<WeakApplication>>(
+        get_comparison_func(),
+        get_can_select_func(tools),
+        get_on_raised_func(tools),
+        get_on_cancelled_func(tools),
+        get_reset_func(),
+        other.window_cycler->get_focus_list(),
+        other.window_cycler->get_originally_focused(),
+        other.window_cycler->get_focused(),
+        other.window_cycler->get_is_active()
+    );
 }
```

Comment:
1. Do we actually copy `ApplicationSelector` instances?
2. `window_cycler` should be initialised, not initialised and then assigned
3. It looks like `WindowCyclingBehavior` needs a copy constructor, not `getters` for all its state

Reply:
I believe `MinimalWindowManager` doesn't have copies deleted, so I needed to define this. But do we actually copy the `MinimalWindowManager`...?